gdnative-derive/src/export.rs

```diff
@@ -0,0 +1,65 @@
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use syn::spanned::Spanned;
+use syn::{DeriveInput, Fields};
+
+fn err_only_supports_fieldless_enums(span: Span) -> syn::Error {
+    syn::Error::new(span, "#[derive(Export)] only supports fieldless enums")
+}
+
+pub(crate) fn derive_export(input: &DeriveInput) -> syn::Result<TokenStream2> {
+    let derived_enum = match &input.data {
+        syn::Data::Enum(data) => data,
+        syn::Data::Struct(data) => {
+            return Err(err_only_supports_fieldless_enums(data.struct_token.span()));
+        }
+        syn::Data::Union(data) => {
+            return Err(err_only_supports_fieldless_enums(data.union_token.span()));
+        }
+    };
+
+    let export_impl = impl_export(&input.ident, derived_enum)?;
+    Ok(export_impl)
+}
+
+fn impl_export(enum_ty: &syn::Ident, data: &syn::DataEnum) -> syn::Result<TokenStream2> {
+    let err = data
+        .variants
+        .iter()
+        .filter(|variant| !matches!(variant.fields, Fields::Unit))
+        .map(|variant| err_only_supports_fieldless_enums(variant.ident.span()))
+        .reduce(|mut acc, err| {
+            acc.combine(err);
+            acc
+        });
+    if let Some(err) = err {
+        return Err(err);
+    }
+
+    let mappings = data
+        .variants
+        .iter()
+        .map(|variant| {
+            let key = &variant.ident;
+            let val = quote! { #enum_ty::#key as i64 };
```

Comment:
It seems like this assumes that the `ToVariant`/`FromVariant` implementations are derived using the official macros with `#[variant(enum = "repr")]`, which isn't necessarily the case. It'd be best if the macro additionally checked for the aforementioned attribute and returned an error if there isn't one, making sure that this assumption holds.

Reply:
Do you mean to check the macro (`#[variant(enum = "repr")]`) is present on the enum, or check the variant conversion is correct? e.g. `#enum_ty::#key.to_variant()` should be integer? which allows users impl their own to/from variant.

Reply:
Getting the value from `to_variant` would be fine as well, but that produces a runtime error, which isn't ideal. It would be useful to have some sort of annotation be the trigger of this mode of the macro anyway, since there are further kinds of `Export` implementations that might be possible to derive (Newtypes maybe?) and thus I'd rather have stricter initial conditions to avoid possible breaking changes.

Reply:
So here are 2 ways to check conversion:

1. check if there is a `#[variant(enum = "repr")]` attribute attached to the derived enum.
2. check if the conversion is correct within `export_info` at runtime, maybe triggered by an attribute e.g. `#[export(runtime_check)]`? I'm not sure if that's an appropriate name.

Reply:
What I meant by "this mode" is the mode of implementing `Export` for enums with all unit variants according to its native representation, so "checking if there is a `#[variant(enum = "repr")]` attribute attached to the derived enum" would be the applicable option out of the two.

The alternative would be checking for something like `#[export(kind = "enum")]`.

Reply:
> The alternative would be checking for something like `#[export(kind = "enum")]`.

I am not sure if I understand, do you mean to use the `kind = ...` to toggle how this macro generates mapping?

Reply:
I've updated to a new version, is the structure what you expected? Currently the parser part is terrible, maybe `ItemAttrBuilder` can be refactored and then used for other macros that need to parse attributes?

https://github.com/godot-rust/gdnative/blob/master/gdnative-derive/src/variant/attr/item.rs#L22

Reply:
Yes, this is roughly what I expected. The "parser" looks fine to me actually. Maybe you can use control structures like `let $pat = $expr else { $diverge };` to flatten the `if-else` hierarchies a bit, but in general I don't think the lack of a common abstraction here is a huge problem. Attribute syntax is very free-form and from my experience there are sometimes cases where the syntax most intuitive to humans turn out to require one-off, not-easily-abstracted logic for machines.

Also, there now exists middleware implementing the parsing abstraction like [`darling`](https://crates.io/crates/darling) that wasn't really a thing back when we started. Depending on the circumstances, it might be better to depend on a library like it than to roll our own. This requires quite a bit of deliberation which I don't think is appropriate to settle within this PR.